File: ts-src/cli/oneshot.ts
import { runAgentLoop } from "../agent/loop.ts";
import type { Provider } from "../provider/types.ts";
import type { SessionOptions } from "../session/setup.ts";
import { createSession } from "../session/setup.ts";
import type { ToolRegistry } from "../tools/registry.ts";
import type { Message } from "../types.ts";

export interface OneshotOptions {
	prompt: string;
	json?: boolean;
	quiet?: boolean;
	agent?: string;
	sessionOptions?: Partial<SessionOptions>;
}

export interface OneshotWithContextOptions {
	prompt: string;
	provider: Provider;
	registry: ToolRegistry;
	messages: Message[];
	json?: boolean;
	quiet?: boolean;
}

export interface OneshotResult {
	output: string;
	exitCode: number;
	toolCalls: number;
}

/**
 * Run a single prompt through the agent loop using a pre-built context.
 * This is the testable core — no session setup, no filesystem side effects.
 */
export async function runOneshotWithContext(options: OneshotWithContextOptions): Promise<OneshotResult> {
	const { prompt, provider, registry, messages } = options;

	if (!prompt) {
		return { output: "No prompt provided", exitCode: 1, toolCalls: 0 };
	}

	// Add the user message
	messages.push({ role: "user", content: prompt });

	let output = "";
	let toolCalls = 0;

	try {
		const gen = runAgentLoop(provider, registry, messages);

		for await (const event of gen) {
			switch (event.type) {
				case "assistant_message": {
					// The last assistant_message with content is the final output
					output = event.message.content ?? "";
					break;
				}
				case "tool_start": {
					toolCalls++;
					break;
				}
				case "error": {
					return {
						output: event.error.message,
						exitCode: 1,
						toolCalls,
					};
				}
			}
		}

		return { output, exitCode: 0, toolCalls };
	} catch (err) {
		const message = err instanceof Error ? err.message : String(err);
		return { output: message, exitCode: 1, toolCalls };
	}
}

/**
 * Run a single prompt end-to-end: create session, execute, return result.
 * This is the CLI entry point — handles session setup and teardown.
 */
export async function runOneshot(options: OneshotOptions): Promise<OneshotResult> {
	if (!options.prompt) {
		return { output: "No prompt provided", exitCode: 1, toolCalls: 0 };
	}

	try {
		const session = await createSession({
			...options.sessionOptions,
			agent: options.agent,
		});

		return runOneshotWithContext({
			prompt: options.prompt,
			provider: session.provider,
			registry: session.registry,
			messages: session.messages,
			json: options.json,
			quiet: options.quiet,
		});
	} catch (err) {
		const message = err instanceof Error ? err.message : String(err);
		return { output: message, exitCode: 1, toolCalls: 0 };
	}
}

/**
 * Format the oneshot result for output to stdout.
 */
export function formatOneshotOutput(result: OneshotResult, options: OneshotOptions): string {
	if (options.json) {
		return JSON.stringify({
			output: result.output,
			exitCode: result.exitCode,
			toolCalls: result.toolCalls,
		});
	}

	return result.output;
}

File: src/cli/oneshot.rs
//! Non-interactive single-prompt execution (for `-p`).

use std::sync::Arc;

use anyhow::Result;
use futures::StreamExt;
use serde_json::json;

use crate::agent::loop_::{run_agent_loop, AgentLoopOptions};
use crate::agent::types::AgentEvent;
use crate::provider::types::Provider;
use crate::session::setup::{create_session, SessionOptions};
use crate::tools::registry::ToolRegistry;
use crate::types::{Message, UserMessage};

#[derive(Debug, Clone, Default)]
pub struct OneshotOptions {
    pub prompt: String,
    pub json: bool,
    pub quiet: bool,
    pub agent: Option<String>,
    pub session_options: Option<SessionOptions>,
}

#[derive(Debug, Clone, Default)]
pub struct OneshotResult {
    pub output: String,
    pub exit_code: i32,
    pub tool_calls: u32,
}

pub async fn run_oneshot_with_context(
    prompt: &str,
    provider: Arc<dyn Provider>,
    registry: ToolRegistry,
    messages: &mut Vec<Message>,
) -> OneshotResult {
    if prompt.is_empty() {
        return OneshotResult {
            output: "No prompt provided".to_string(),
            exit_code: 1,
            tool_calls: 0,
        };
    }
    messages.push(Message::User(UserMessage {
        content: prompt.to_string(),
    }));

    let mut output = String::new();
    let mut tool_calls = 0u32;
    let mut stream = run_agent_loop(provider, registry, messages, AgentLoopOptions::default());
    while let Some(event) = stream.next().await {
        match event {
            AgentEvent::AssistantMessage { message, .. } => {
                output = message.content.unwrap_or_default();
            }
            AgentEvent::ToolStart { .. } => tool_calls += 1,
            AgentEvent::Error { message } => {
                return OneshotResult {
                    output: message,
                    exit_code: 1,
                    tool_calls,
                }
            }
            _ => {}
        }
    }
    OneshotResult {
        output,
        exit_code: 0,
        tool_calls,
    }
}

pub async fn run_oneshot(options: OneshotOptions) -> OneshotResult {
    if options.prompt.is_empty() {
        return OneshotResult {
            output: "No prompt provided".to_string(),
            exit_code: 1,
            tool_calls: 0,
        };
    }
    let mut session_opts = options.session_options.clone().unwrap_or_default();
    if options.agent.is_some() {
        session_opts.agent = options.agent.clone();
    }
    let session = match create_session(session_opts).await {
        Ok(s) => s,
        Err(e) => {
            return OneshotResult {
                output: e.to_string(),
                exit_code: 1,
                tool_calls: 0,
            }
        }
    };
    let mut messages = session.messages;
    run_oneshot_with_context(
        &options.prompt,
        session.provider,
        session.registry,
        &mut messages,
    )
    .await
}

pub fn format_oneshot_output(result: &OneshotResult, options: &OneshotOptions) -> String {
    if options.json {
        return json!({
            "output": result.output,
            "exitCode": result.exit_code,
            "toolCalls": result.tool_calls,
        })
        .to_string();
    }
    result.output.clone()
}

#[allow(dead_code)]
fn _r() -> Result<()> {
    Ok(())
}
